ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/common/TransactionHelperImpl.java

```diff
@@ -0,0 +1,103 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.narayana.jta.jms.common;
+
+import javax.jms.JMSException;
+import javax.transaction.Status;
+import javax.transaction.Synchronization;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import javax.transaction.xa.XAResource;
+
+/**
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+public class TransactionHelperImpl implements TransactionHelper {
+
+    private final TransactionManager transactionManager;
+
+    public TransactionHelperImpl(TransactionManager transactionManager) {
+        this.transactionManager = transactionManager;
+    }
+
+    @Override
+    public Transaction getTransaction() throws JMSException {
+        try {
+            return transactionManager.getTransaction();
+        } catch (SystemException e) {
+            throw new JMSException("Failed to get transaction: "
+                    + e.getMessage());
+        }
+    }
+
+    @Override
+    public boolean isTransactionAvailable() throws JMSException {
+        try {
+            return transactionManager.getStatus() != Status.STATUS_NO_TRANSACTION;
+        } catch (SystemException e) {
+            throw new JMSException("Failed to get transaction status: "
+                    + e.getMessage());
+        }
+    }
+
+    @Override
+    public boolean isTransactionActive() throws JMSException {
+        try {
+            return transactionManager.getStatus() == Status.STATUS_ACTIVE;
+        } catch (SystemException e) {
+            throw new JMSException("Failed to get transaction status: "
+                    + e.getMessage());
+        }
+    }
+
+    @Override
+    public void registerSynchronization(Synchronization synchronization) throws JMSException {
+        try {
+            getTransaction().registerSynchronization(synchronization);
+        } catch (Exception e) {
+            throw new JMSException("Failed to register synchronization: "
+                    + e.getMessage());
+        }
+    }
+
+    @Override
+    public void enlistResource(XAResource xaResource) throws JMSException {
+        try {
+            getTransaction().enlistResource(xaResource);
+        } catch (Exception e) {
+            throw new JMSException("Failed to enlist XA resource: "
+                    + e.getMessage());
+        }
+    }
+
+    @Override
+    public void delistResource(XAResource xaResource) throws JMSException {
+        try {
+            getTransaction().delistResource(xaResource, XAResource.TMSUCCESS);
+        } catch (SystemException e) {
+            throw new JMSException("Failed to delist XA resource: "
```

Comment:
Is it possible to set the e as the root cause on the thrown JMSException so you don't lose the stack trace